pkg/kubelet/dockertools/exec.go

```diff
@@ -90,15 +88,22 @@ func (*NsenterExecHandler) ExecInContainer(client DockerInterface, container *do
 		if stderr != nil {
 			command.Stderr = stderr
 		}
-
-		return command.Run()
+		if err := command.Start(); err != nil {
+			return err
+		}
+	}
+	if timeout > 0 {
+		defer time.AfterFunc(timeout, func() {
```

Comment:
@vishh https://golang.org/pkg/time/#AfterFunc
`AfterFunc` runs the function in it's own goroutine, so we don't need the `go` keyword to create yet another one.

The `defer` here is deferring a call to `Stop`, the `timer` (returned by `AfterFunc`) method to abort the execution.

Basically it means "if it takes more than timeout, kill the process, otherwise, forget about killing it".


We only need to expand this into a pattern similar to what you posted if we want to return an explicit timeout error instead of what `Wait` returns when the process is killed.

I'm happy with getting the error that says the process was killed, but have nothing against doing differently if preferred.